### DISCUSSION TITLE
Flow of the global route middleware in Nuxt 3

### DISCUSSION THREAD

**Discussion by AloisSeckar** [nuxt/nuxt#19361] [Questions]
Hi, I have a couple of questions regarding middleware, I'd like to clarify, because I wasn't able to confirm myself just from the docs:

1. Is there a guaranted order of execution for global `defineNuxtRouteMiddleware`? From my testing it seems, that it is always alphabetical based on file name, but without knowing for sure it might be just a coincidence.
2. When `return navigateTo('/foo')` is called, are the remaining middlewares cancelled or do they run anyway?
3. And the last question, that can as well be asked first - is it even a good idea to have more global route guards performance wise? Or is it better to put all logic into just one file?

**Comment by viii47** [comment]
> 3. And the last question, that can as well be asked first - is it even a good idea to have more global route guards performance wise? Or is it better to put all logic into just one file?

This is more of a *project architecture design* question and how much *scalability* you need to know how to approach it. The performance should not perforate or be inherently worse with multiple global route guards - it depends more how they're designed and what routine they run. 

I personally do not recommend to have *everything* in one certain file *if* the logic differs too much, but that's again a question you need to ask yourself how much that project is able to grow over time. If it's a small project, putting all logic into one file might be easier to maintain over time if only life support is necessary. I've had some projects (not in Nuxt but Angular - same logic applies though) that grew logically so much that having split up middleware files (global & named) made more sense.

**Reply by AloisSeckar** [reply]
Well, should be really big and long living project. And therefore I would indeed tend up to split the logic into more files. But since I haven't grasped much about the things running under hood, I wanted to make sure, that the overhead of setting up/tearing down middlewares isn't too big to handle as their numbers grow. Or what is the reasonable limit.
